### PR TITLE
[receiver/skywalking] Fix skywalking traceid and spanid convertion.

### PR DESCRIPTION
**Pull request by taloric**
**Description:**
- Backgrounds:
  - [Skywalking](https://skywalking.apache.org/) use `segments` to cover spans as a group. Also `SegmentId` + `SpanId` to identify an unique span in `TraceData`.
  -  Protocols could be found [here](https://github.com/apache/skywalking/blob/master/docs/en/protocols/Trace-Data-Protocol-v3.md).
  - The Skywalking receive server handle one segment object in each http request. Those segments have different `segmentId` and share the same `traceId`.

- Code Changes:
  - Use a new rule to convert `Skywalking-TraceId` to `Opentelemetry-TraceId`. (The old rules breaks relations between segments because it use pointer of strings to generate a new `traceId`.)
    - Skywalking-TraceId format be like: `de5980b8-fce3-4a37-aab9-b4ac3af7eedd`([RFC4122](https://datatracker.ietf.org/doc/html/rfc4122)) or `56a5e1c519ae4c76a2b8b11d92cead7f.1.16563474296430001`.
  - Use `SegmentId` in the convertion of `SpanId` to make sure the `SpanId` is globally unique.
    - `SegmentId` format be like: `4f2f27748b8e44ecaf18fe0347194e86.33.16560607369950066`.
    - The first part of `SegmentId`(`4f2f27748b8e44ecaf18fe0347194e86`) will repeat multiple times in a `TraceData`, so the last part is the real unique identifier of segments.

**Testing:** 
- Add `Test_stringToTraceID()`: Make sure Skywalking-TraceId convert to the Otel-TraceId by a rule.
- Add `Test_segmentIdToSpanId()`: Using Skywalking segmentId & spanId to generate a new Otel-SpanId.

**Documentation:** 
No changes.

### CODE REVIEW

**Comment by taloric** [comment]
- Add Jaeger test result here (with otel agent) :
skywalking-jaeger

- Same application trace in Skywalking dashboard (with skywalking agent) : 
skywalking-form2

**Review by codeboten — COMMENTED** [review]
Pinging code owner @JaredTan95, please review

File: receiver/skywalkingreceiver/skywalkingproto_to_traces.go [hunk 1]
```diff
@@ -236,37 +238,68 @@ func microsecondsToTimestamp(ms int64) pcommon.Timestamp {
 }
 
 func stringToTraceID(traceID string) pcommon.TraceID {
-	return pcommon.NewTraceID(unsafeStringToBytes(traceID))
+	dst, err := stringTo16Bytes(&traceID)
+	if err != nil {
+		return pcommon.InvalidTraceID()
+	}
```
Comment (sharang) [inline receiver/skywalkingreceiver/skywalkingproto_to_traces.go:244]: use https://pkg.go.dev/github.com/google/UUID#Parse

Comment (taloric) [reply receiver/skywalkingreceiver/skywalkingproto_to_traces.go:245]: Done. Now Jaeger bring more informations:
image

**Comment by taloric** [comment]
@JaredTan95 
I found that Skywalking `traceid` has 2 formats , like `de5980b8-fce3-4a37-aab9-b4ac3af7eedd` or `56a5e1c519ae4c76a2b8b11d92cead7f.1.16563474296430001` ( [Skywalking Live Demo](http://demo.skywalking.apache.org/general) ), but even in [protocols](https://skywalking.apache.org/docs/main/latest/en/protocols/trace-data-protocol-v3/) I can find no descriptions about that . Wondering if you have any idea of this...

**Comment by JaredTan95** [comment]
> * Using Skywalking segmentId & spanId to generate a new Otel-Spa

`de5980b8-fce3-4a37-aab9-b4ac3af7eedd` format is produced by the browser/js SDK or envoy/nginx-lua sdk.

**Comment by taloric** [comment]
> 
> hi, @taloric Thanks for the fix, I don't see that your screenshot involves using the `skywalking agent` to send tracing data to `otel collector skywalking receiver` and exporter into `jaeger`. Can you update one?

Thanks for the review. I did find some bugs during test with `skywalking agent`. I make some changes about `traceid` convertion rules.

- Here are my test result (send `trace data` with [skywalking-java-agent](https://github.com/apache/skywalking-java))
- They build the same results: 
- Jaeger:
jaeger-java-agent-test

- skywalking: 
skywalking-java-agent

---

- Also, I make another test with [skywalking-python-agent](https://github.com/apache/skywalking-python). It brings the same trace tree.
- Jaeger:
python-agent-jaeger2
- skywalking:
skywalking-python-agent

**Review by JaredTan95 — APPROVED** [review]
good catches, LGTM
@taloric BTW,  can you change the issue title to `[receiver/skywalking] Fix skywalking traceid and spanid convertion.` ?

**Comment by JaredTan95** [comment]
@codeboten  PTAL on behalf of approvers.

**Review by sharang — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -38,6 +38,7 @@
 - `sapmreceiver`: Fix issue where component instance use in multiple pipelines leads to start failures (#11518)
 - `signalfxreceiver`: Fix issue where component instance use in multiple pipelines leads to start failures (#11513)
 - `splunkhecreceiver`: Fix issue where component instance use in multiple pipelines leads to start failures (#11517)
+- `skywalkingreceiver`: Fix skywalking traceid and spanid convertion (#11562)
```
Comment (JaredTan95) [inline CHANGELOG.md:41]: Please use the new way of adding changelog entries
👉 https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-a-changelog-entry

**Review by JaredTan95 — APPROVED** [review]
LGTM
thx for your contributions

**Comment by JaredTan95** [comment]
@Aneurysm9 PTAL on behalf of approvers.

**Comment by taloric** [comment]
Skywalking Agent will write skywalking trace info in logs like this:
image
But after conversion, the relationship between `trace` and `log` were lost. So I add skywalking source information in attributes. 

@Aneurysm9 @codeboten @JaredTan95  please review.

**Comment by JaredTan95** [comment]
> So I add skywalking source

This is another bug or caused by your pr?

**Comment by taloric** [comment]
> > So I add skywalking source
> 
> This is another bug or caused by your pr?

It's not caused by this PR. And It's aim to solve the same issue.

I think what our expectation is: `application -> Trace data -> otel / skywalking` is a full `Trace`. We can track any step in this `Trace` with `traceid` or `spanid`. 

This PR aim to solve `Trace` broken issues when skywalking `traceid` convert to otel `traceid`. (which was done in the first commit. ) But that didn't build relations between application `Trace` logs and otel `Trace` export result.

So I add skywalking original `traceid` as attributes to connect application logs with otel export results. Now you can get any skywalking `traceid` in application logs, and find the same `Trace` directly in Jaeger or other backends.

**Comment by JaredTan95** [comment]
> 

got it. make sense, LGTM

**Comment by taloric** [comment]
@codeboten @Aneurysm9 @bogdandrutu would you please take a look at this PR?

**Review by codeboten — APPROVED** [review]

**Pull request merged**